### ISSUE TITLE
VSC preview in 2.9.3 is not working as in 2.8.10

### ISSUE THREAD

**Issue by ioannisstavrakakis** [asciidoctor/asciidoctor-vscode#501]
## Description

The VSC preview in v2.9.3 is not working the same way as in v2.8.10.
I use the option:
**Asciidoc › Preview: Use Editor Style**
_Use editor style instead of default asciidoctor.css_

The problem is also seen in v2.9.2, so it is something between this and v2.8.10 that broke the preview.

There are so many changes between the two previews that it is best if you check the attached gif files to see what exactly has been changed.

## System Information

Visual Studio Code:
Version: 1.63.2 (user setup)
Commit: 899d46d82c4c95423fb7e10e68eba52050e30ba3
Date: 2021-12-15T09:40:02.816Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19044

asciidoctor-vscode v2.9.3

## To Reproduce

Steps to reproduce the issue:

1. Go to extension settings.
2. Go to **Asciidoc › Preview: Use Editor Style**
3. Select the checkbox _Use editor style instead of default asciidoctor.css_
4. Check adoc preview pane.

### 2.9.3. preview
![2 9 3](https://user-images.githubusercontent.com/87022642/152146932-5c84ebc9-22dd-4a05-8d1b-c95d21f7b6bc.gif)

### 2.8.10 preview
![2 8 10](https://user-images.githubusercontent.com/87022642/152147574-b1147ee8-2363-4355-a226-38f81678465b.gif)

**Comment by Mogztter** [comment]
Could you please share your document?

**Comment by ioannisstavrakakis** [comment]
It is just a doc with various adoc syntaxes for reference. There you go:
[asciidoc_syntax_reference_guide.zip](https://github.com/asciidoctor/asciidoctor-vscode/files/7986413/asciidoc_syntax_reference_guide.zip)

**Comment by Mogztter** [comment]
I cannot reproduce this issue:

![image](https://user-images.githubusercontent.com/333276/152176470-fec9166a-156f-4362-94e8-142aa5baa592.png)


```
Version: 1.63.2
Commit: 899d46d82c4c95423fb7e10e68eba52050e30ba3
Date: 2021-12-15T09:39:46.686Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.4.0-96-generic
```

**Comment by ioannisstavrakakis** [comment]
The thing is that all writers in my team have the same issue, so it is not just me :)
This also happens irrespective of the VSC color theme, any dark or light one.

I can post you the result of the console, if it is going to help you out:
[vscode-app-1643873041850.log](https://github.com/asciidoctor/asciidoctor-vscode/files/7992545/vscode-app-1643873041850.log)
[vscode-app-1643873047190.log](https://github.com/asciidoctor/asciidoctor-vscode/files/7992546/vscode-app-1643873047190.log)

**Comment by ioannisstavrakakis** [comment]
I checked in Elements and there are some differences in the following element, possibly affecting the preview:

![image](https://user-images.githubusercontent.com/87022642/152300925-be6875eb-c360-44f2-bbce-af9dfbcd9ca7.png)

I am also attaching the values of the same element between 2.9.3 and 2.8.10:
[vsc.xlsx](https://github.com/asciidoctor/asciidoctor-vscode/files/7992644/vsc.xlsx)

**Comment by Mogztter** [comment]
> I can post you the result of the console, if it is going to help you out:

Nothing alarming in the logs.

> I checked in Elements and there are some differences in the following element, possibly affecting the preview:

Not sure, it seems that you are missing the default Asciidoctor stylesheet, I can get a similar output when I manually remove the `<style>` element that contains the default stylesheet. Are you using a custom stylesheet? Do you extend the default stylesheet?

If you inspect the HTML, you should find a `<style>` element just after the document title that contains the default stylesheet.
Here's a trimmed down example of what you should get:

```html
<html lang="en" style="">
<head>
    <style id="_defaultStyles">
        /* ... */
    </style>
    <script id="_vscodeApiScript">
      /* ... */
    </script>
    <meta http-equiv="Content-type" content="text/html;charset=UTF-8">
    <meta id="vscode-asciidoc-preview-data" data-settings="{}">
    <script src="https://path/to/dist/pre.js" nonce="1643878011522522"></script>
    <style nonce="1643878011522522">
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', 'HelveticaNeue-Light', 'Ubuntu', 'Droid Sans', sans-serif;
            font-size: 14px;
            line-height: 1.6;
        }
    </style>
    <base href="https://path/to/file.adoc">
</head>
<body class="article data-line-1 vscode-body scrollBeyondLastLine wordWrap showEditorSelection vscode-dark" role="document" data-vscode-theme-kind="vscode-dark" data-vscode-theme-name="Dark+ (default dark)">
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.16">
<meta name="keywords" content="AsciiDoc, Syntax,  Reference, Guide, technical writing">
<meta name="author" content="...">
<title>Netcompany-Intrasoft: AsciiDoc Syntax Reference Guide</title>
<style>
    /* Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
   /* content... 🎯 */
</style>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link rel="stylesheet" href="https://path/to/media/highlightjs/styles/github.min.css">
<div id="header">
    <h1>Netcompany-Intrasoft: AsciiDoc Syntax Reference Guide</h1>
    <!-- content... -->
</div>
</body>
</html>
```

**Comment by danyill** [comment]
I had a look too. Using the document provided and extension version 2.9.3 in the deskop app on Linux:

* Using VS Code 1.61.2 I cannot reproduce this.
* Using VS Code 1.63.2 I cannot reproduce this.

I haven't seen any errors in the extension log.
I tried different security settings and also saw no incorrect behaviour.

Could you try disabling all other extensions, restarting VS Code and seeing if the problem persists?
Do you have any unusual settings in a .vscode or user settings file?

Otherwise I'd suggest removing all styles and all document header attributes if they aren't critical to reproducing the error. If the problem still persists, retain the "flawed output" and cut document text out before and after. If you're left only with the "flawed output" and nothing else in the document then we must try to investigate further.

**Comment by ioannisstavrakakis** [comment]
> Could you try disabling all other extensions, restarting VS Code and seeing if the problem persists?
Do you have any unusual settings in a .vscode or user settings file?
I uninstalled all extensions except this one, and the problem still exists in 2.9.3.

I am not using a custom stylesheet, not extend the default one.

However, the <style> element is empty when I update to 2.9.3, while it is fully completed when I downgrade to 2.8.10.
So copying and pasting the same values does the trick and preview works (except the borders of tables and cells) , but question is why this element gets reset?

**Comment by ioannisstavrakakis** [comment]
> Otherwise I'd suggest removing all styles and all document header attributes if they aren't critical to reproducing the error. If the problem still persists, retain the "flawed output" and cut document text out before and after. If you're left only with the "flawed output" and nothing else in the document then we must try to investigate further.

I did so, the problem still persists. I even created a new adoc with just a title, pone heading and one paragraph, no atributes, still the same issue.
![image](https://user-images.githubusercontent.com/87022642/152318776-73b1f864-52f1-4d19-9881-0f81cfd69d1e.png)

**Comment by danyill** [comment]
Recognising that you have this ticked at the moment and want it to stay like this, if you untick it, does it render correctly with the Asciidoctor stylesheet?

i.e. Untick the following:

![image](https://user-images.githubusercontent.com/674783/152319106-ea216c99-90cc-48c1-a78d-cb17ee3c6b70.png)

**Comment by ioannisstavrakakis** [comment]
Yes it does, but only in white style..

![image](https://user-images.githubusercontent.com/87022642/152320823-2d7852ab-1f5f-42a4-b160-de203687f765.png)

and it is kind of disturbing to have black editor and white preview..

**Comment by ioannisstavrakakis** [comment]
Is there any way I can troubleshoot this otherwise? Or any file I can just take from the 2.8.10 installation folder and replace in the 2.9.3 one?

**Comment by Mogztter** [comment]
You should try to completely uninstall VS code and remove all associated files (`.vscode`...) from your machine.

Then, install the latest version of VS code again and install the extension.

Are you using VS code from a local file system?

**Comment by Mogztter** [comment]
Might be a Windows related issue, I will try to reproduce using Windows.

**Comment by ioannisstavrakakis** [comment]
> You should try to completely uninstall VS code and remove all associated files (`.vscode`...) from your machine.
> 
> Then, install the latest version of VS code again and install the extension.
> 
> Are you using VS code from a local file system?

I uninstalled VSC, removed all folders found, `.vscode, Code, Microsoft VS Code`, did clean installation of VSC, no previous settings detected, installed only the asciidoctor extension, same behavior.

I repeated this for VSC 1.64 which showed up today, same result.

So there is definitely something wrong with the extension in Windows.

**Comment by danyill** [comment]
I've been able to replicate this behaviour on a Windows machine using the sample provided:

```asciidoc
= eeee

== title
this


[cols="4*"]
|===

|www|www|ww|www|
|rrr|rrrr|rr|rrrr

|===

wef
wefw e
```

I don't see any extension host errors or relevant console log errors. I'll keep looking.

**Comment by danyill** [comment]
Our development is rather heavily focussed on Linux and our npm scripts don't work well on Windows. After working around this, I find the problem is occurring associated with the update of path locations to URI-s for the web version.

If I understand correctly, in 5715c52f84e51c19a04f8ab890b51de03f88b6f6 we did [this](https://github.com/asciidoctor/asciidoctor-vscode/commit/5715c52f84e51c19a04f8ab890b51de03f88b6f6#diff-9e0ba4783c51e8ac43c4c4332257d2b9aeee7c18c1209d7f90390a467f2e7781L18-R23):

```diff
-private extPath = vscode.extensions.getExtension('asciidoctor.asciidoctor-vscode').extensionPath
-private stylesdir = path.join(this.extPath, 'media')
+private stylesdir: string

-constructor (private readonly filename: string, private errorCollection: vscode.DiagnosticCollection = null) { }
+ constructor (private readonly filename: string, private errorCollection: vscode.DiagnosticCollection = null) {
+   const extensionContext = vscode.extensions.getExtension('asciidoctor.asciidoctor-vscode')
+   this.stylesdir = vscode.Uri.joinPath(extensionContext.extensionUri, 'media').toString()
+  }
```

When I place a debug point  where we call asciidoctor.js:

https://github.com/asciidoctor/asciidoctor-vscode/blob/1d96cb6d49365c07ed6f22fb8b6af70acdc5d6af/src/asciidocParser.ts#L152

I note what I expect for `options.attributes.stylesdir` is something like:

    c:\Users\zero1\Documents\asciidoctor-vscode-l\asciidoctor-vscode\media

But what I currently get with our changes above is:

    file:///c%3A/Users/zero1/Documents/asciidoctor-vscode-l/asciidoctor-vscode/media

This is used directly as the `stylesdir` attribute in the `options` object and this means the directory is not correctly resolved to locate stylesheet resources.

This problem doesn't occur on Linux (better path handling :wink:) but only on Windows 

Guillaume has previously requested we refactor `asciidocParser.ts` and to be fair the logic is very hard to follow and munges together many concerns.

For reference, the area where I need to go through the path handling is here:

https://github.com/asciidoctor/asciidoctor-vscode/blob/1d96cb6d49365c07ed6f22fb8b6af70acdc5d6af/src/asciidocParser.ts#L20-L152

I'm not sure of the solution yet but I hope to work on this a little later this weekend.

**Comment by danyill** [comment]
I thought the solution could be as simple as instead of using:

```diff
-this.stylesdir = vscode.Uri.joinPath(extensionContext.extensionUri, 'media').toString()
+this.stylesdir = vscode.Uri.joinPath(extensionContext.extensionUri, 'media').path
```

However it was not. Calling `.path` does not yield a Windows path, instead it yield:

    /c:/Users/zero1/Documents/asciidoctor-vscode-l/asciidoctor-vscode/media

However calling `.fsPath` seems to be a stable reference and I have tested on Linux/Windows but breaks the web version. I think we need to test and set differently for this case.

On Windows `.fsPath` gives:

    c:/Users/zero1/Documents/asciidoctor-vscode-l/asciidoctor-vscode/media

Hopefully, a PR shortly.

**Comment by Mogztter** [comment]
A browser is capable of resolving: `file:///c%3A/Users/zero1/Documents/asciidoctor-vscode-l/asciidoctor-vscode/media`. 
However, I think Asciidoctor (or at least Asciidoctor.js) expects `styledir` to be a path (not a URL).
Not sure how/why it's working on Linux, I will need to do some tests on Asciidoctor.js.

I'm pretty sure it's working in a browser environment, because Asciidoctor.js is using `XMLHttpRequest` to get the stylesheet from a URL (since we are running in a browser environment).

I will open a discussion at https://asciidoctor.zulipchat.com/ so we can decide what is the path forward. I don't mind fixing this issue with a "workaround" but I think it's important that we decide how we want to solve this issue (because we keep hitting a wall).

**Comment by ioannisstavrakakis** [comment]
Thanks for the heads-up! Looks pretty complicated after all :/

**Comment by danyill** [comment]
> Not sure how/why it's working on Linux, I will need to do some tests on Asciidoctor.js.

I can confirm that if `allow-uri-read` is set on Linux then setting `stylesdir` to:

    file:///home/mulhollandd/Documents/asciidoc_testing/x

or

    /home/mulhollandd/Documents/asciidoc_testing/x

will resolve a stylesheet reference correctly. Depending on what we're trying to do (see Zulip) the Windows behaviour could be seen as a bug. Discussions here:

* https://asciidoctor.zulipchat.com/#narrow/stream/279642-users/topic/Support.20local.20URI.20as.20stylesdir
* https://asciidoctor.zulipchat.com/#narrow/stream/279642-users/topic/Path.20handling.20in.20Asciidoctor

**Comment by Mogztter** [comment]
Fixed in https://github.com/asciidoctor/asciidoctor-vscode/commit/35f2cc1f5ff27ad54ac6064691f5d7ac57386988

**Comment by ioannisstavrakakis** [comment]
Works like a charm! Many thanks!!!
